blog/_posts/2017-01-24-cudanative-nvidia.md

```diff
@@ -0,0 +1,233 @@
+of two vectors:
+
+```julia
+using CUDAdrv, CUDAnative
+
+function kernel_vadd(a, b, c)
+    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
+    c[i] = a[i] + b[i]
+
+    return nothing
+end
+
+dev = CuDevice(0)
+ctx = CuContext(dev)
+
+# generate some data
+len = 512
+a = rand(Int, len)
+b = rand(Int, len)
+
+# allocate & upload on the GPU
+d_a = CuArray(a)
+d_b = CuArray(b)
+d_c = similar(d_a)
+
+# execute and fetch results
+@cuda (1,len) kernel_vadd(d_a, d_b, d_c)
+c = Array(d_c)
+
+destroy(ctx)
+```
+
+The real workhorse of this example is the `@cuda` macro, which generates specialized code
+for compiling the kernel function to GPU assembly, uploading it to the driver, and preparing
+the execution environment. Together with Julia's JIT compiler, this results in a very
+efficient kernel launch sequence, avoiding the runtime overhead typically associated with
+dynamic languages. The generated code is also specific to the types of the arguments;
+Calling the same kernel with differently typed arguments would Just Work.
+
+<!-- Mention libNVVM? -->
+
+Behind the scenes, CUDAnative.jl hooks into the Julia compiler to lower kernel functions to
+LLVM IR, an intermediate representation used by the LLVM compiler framework. We've created
+[LLVM.jl](https://github.com/maleadt/LLVM.jl) to interface with the framework, and use it to
+compile IR to PTX code, the virtual instruction set used to program CUDA GPUs. Finally, we
+use CUDAdrv.jl to interact with the CUDA driver: upload the PTX code, compile it to
+hardware-specific Shader Assembly (SASS), and launch the kernel.
+
+![Overview of the different components making up the native programming
+support.](/images/blog/2017-01-24-cudanative/overview.png)
+
+By reusing the Julia compiler to generate IR we avoid multiple language implementations with
+slightly different semantics. That means you can use all features of the Julia language,
+like metaprogramming or dynamically-typed multimethods, to implement GPU functions. There
+are still [some limitations]({{page.cudanative_tree}}#julia-support), for example we haven't
+ported Julia's runtime library or implemented a GPU garbage-collector yet, but these aren't
+structural limitations.
+
+
+### Example: parallel reduction
+
+For a more realistic example, we'll have a look at a parallel reduction [as documented on
+this blog
+before](https://devblogs.nvidia.com/parallelforall/faster-parallel-reductions-kepler/). The
+Julia version is part of the [CUDAnative
+examples]({{page.cudanative_tree}}/examples/reduce/reduce.jl) and too long to discuss in
+full, so let's zoom in on some details instead. Listing 2 shows the innermost GPU function
+responsible for reducing within a warp using shuffle instructions (slightly simplified for
+readability's sake).
+
+```julia
+function reduce_warp(op, val)
+    offset = CUDAnative.warpsize() ÷ 2
+    while offset > 0
+        val = op(val, shfl_down(val, offset))
+        offset ÷= 2
```

Comment:
Someone reading this who is not familiar with Julia might wonder why you're using `÷` instead of `/`.  Might be worth a brief comment (either inline or below).